src/utils/allocators.zig

```diff
@@ -2,6 +2,269 @@ const std = @import("std");
+/// If the allocation size being too to be recycled after trying to collapse smaller records
+/// (config.max_collapse_tries times) then we panic (or return error.CollapseFailed in allocUnsafe).
+pub fn RecycleBuffer(comptime T: type, default_init: T, config: struct {
+    /// If enabled, all operations will require an exclusive lock.
+    thread_safe: bool = !builtin.single_threaded,
+    max_collapse_tries: u32 = 5,
+    collapse_sleep_ms: u32 = 100,
+    min_split_size: u64 = 128,
+}) type {
+    std.debug.assert(config.min_split_size > 0);
+
+    return struct {
+        // records are used to keep track of the memory blocks
+        records: std.ArrayList(Record),
+        // memory holds blocks of memory ([]T) that can be allocated/deallocated
+        memory: std.ArrayList([]T),
+        // allocator used to alloc the memory blocks
+        memory_allocator: std.mem.Allocator,
+        // total number of T elements we have in memory
+        capacity: u64,
+        // for thread safety
+        mux: std.Thread.Mutex = .{},
+
+        // NOTE: we use the global_index to support fast loading the state
+        pub const Record = struct {
+            is_free: bool,
+            buf: []T,
+            global_index: u64,
+            len: u64,
+            // NOTE: this is tracked for correct usage of collapse()
+            memory_index: u64,
+
+            /// NOTE: we dont want to re-write the buf memory (that exists in `memory` field) on save
+            pub const @"!bincode-config:buf": sig.bincode.FieldConfig([]T) = .{
+                .skip = true,
+                .default_value = &.{},
+            };
+        };
+        const Self = @This();
+
+        const AllocatorConfig = struct {
+            memory_allocator: std.mem.Allocator,
+            records_allocator: std.mem.Allocator,
+        };
+
+        pub fn init(allocator_config: AllocatorConfig) Self {
+            return .{
+                .records = std.ArrayList(Record).init(allocator_config.records_allocator),
+                .memory = std.ArrayList([]T).init(allocator_config.records_allocator),
+                .memory_allocator = allocator_config.memory_allocator,
+                .capacity = 0,
+            };
+        }
+
+        pub fn create(allocator_config: AllocatorConfig) !*Self {
+            const self = try allocator_config.records_allocator.create(Self);
+            self.* = Self.init(allocator_config);
+            return self;
+        }
+
+        pub fn deinit(self: *Self) void {
+            if (config.thread_safe) self.mux.lock();
+            defer if (config.thread_safe) self.mux.unlock();
+
+            for (self.memory.items) |block| {
+                self.memory_allocator.free(block);
+            }
+            self.memory.deinit();
+            self.records.deinit();
+        }
+
+        /// append a block of N elements to the manager
+        pub fn append(self: *Self, n: u64) std.mem.Allocator.Error!void {
```

Comment:
nit: `append` is mostly used to add actual elements. This seems to be expanding the capacity of the memory that can be later allocated. Maybe similar to zig's std have name that include `ensure*capacity` or `expand*capacity`